chore: reinstating tubular to prod-ready for retirement

deborahgu opened the pull request:
these user retirement scripts haven't been updated since the re-factor back into  edx-platform a year ago. This is catching up with all of the interim changes. It's not just a mass copy, for two reasons: the file paths and package paths for imports are different; and this is configured as a solo package with a setup.py.

the changes include:

* the updates in removing _some_ ancient functionality and workarounds specific to old versions of Python.
* Switching to the  included version of `mock`, per above
* Removing all code related to the Demographics pipeline
* updating or removing pip constraints
* ran `make upgrade`  to get newer pip files.
* a re-factor that was apparently necessary for `moto` updates. I believe this is a  package that allows mocking the AWS `boto` library.
* updating the package to run Python 3.11 on ubuntu-latest  (in github workflows, in `setup.py`, in the `Dockerfile`, etc.)

This does not include any formatting updates, and it does not add any changes that were not already added to edx-platform. So, for example, there are some ancient uses of `from six import ...`  that have not been removed. There are also some unordered imports, and other ancient functionality. The goal here is to just get this matching the version on platform as much as possible.

FIXES: APER-4019

doctoryes commented:
👍

deborahgu committed:
chore: reinstating tubular to prod-ready for retirement

these user retirement scripts haven't been updated since the re-factor
back into  edx-platform a year ago. This is catching up with all of the
interim changes. It's not just a mass copy, for two reasons: the file
paths and package paths for imports are different; and this is
configured as a solo package with a setup.py.

the changes include:

* the updates in removing _some_ ancient functionality and workarounds specific to old versions of Python.
* Switching to the  included version of `mock`, per above
* Removing all code related to the Demographics pipeline
* updating or removing pip constraints
* ran `make upgrade`  to get newer pip files.
* a re-factor that was apparently necessary for `moto` updates. I believe this is a  package that allows mocking the AWS `boto` library.
* updating the package to run Python 3.11 on ubuntu-latest  (in github workflows, in `setup.py`, in the `Dockerfile`, etc.)

This does not include any formatting updates, and it does not add any changes that were not already added to edx-platform. So, for example, there are some ancient uses of `from six import ...`  that have not been removed. There are also some unordered imports, and other ancient functionality. The goal here is to just get this matching the version on platform as much as possible.

FIXES:
* master

deborahgu committed:
chore: modified non-retirement scripts  for package update

the moto update also affected some nonretirement scripts. I don't know
if these are used in production at all anymore, but I obviously chose
to make the same re-factor for those scripts, since I still updated the
package.

deborahgu committed:
chore:  missing import

adding a missing import from the mock import modification

justinhynes commented on Dockerfile:
```diff
@@ -1,4 +1,6 @@
-FROM python:3.8-buster
+FROM ubuntu:jammy AS base
+
+ARG PYTHON_VERSION=3.11
```
I've never seen this image used before in other Dockerfiles, so I was digging a bit deeper.

So my understanding is that the `python:3.8-buster` was a Debian-based image on the `buster` release with Python 3.8 preinstalled. If we move to the `ubuntu:jammy` base image, I don't believe it will have Python installed. We will have to do a little bit of work to ensure that the version of Python we want is installed.

Is there an official Python 3.11 image that we can use?

justinhynes replied:
I think this is why we do things like this in our IDA Dockerfiles: https://github.com/edx/public-dockerfiles/blob/cb21c839eb1f1d48b3ad2285c0c4ef3a89d3398f/dockerfiles/credentials.Dockerfile#L8C1-L41C14

deborahgu replied:
ah got it, I copied from that dockerfile but just switched to jammy because CI used `ubuntu-latest`. I'm happy to switch to focal and then add the apt commands. 👍

justinhynes commented on README.md:
```diff
@@ -29,11 +32,9 @@ and migrated [here](https://github.com/openedx/edx-platform/tree/master/scripts/
+### 2U Fork
 
-**2U Fork:** After being deprecated for Open edX, this repo has been forked into the `edx` GitHub org and is maintained by (and for) 2U.
+After being deprecated for Open edX, this repo was forked into the `edx` GitHub org and is maintained by (and for) 2U. Initially, on 13 March, 2024 2U switched from `tubular` to `edx-platform` versions of the retirement scripts, but in May 2025, 2U  temporarily switched back to running user from this repository,, in order to most efficiently solve the problem of divergent third-party retirement s ...
```
nit: double comma

justinhynes reviewed (changes requested):
Most of this LGTM, but I'm blocking this based on the comment I left on the Dockerfile.

deborahgu committed:
chore: new dockerfile

Changes from code review:

* cleaning up one typo
* looks like the only way to get a docker  image that combines
  ubuntu-focal and 3.11  (which I'm going with because we have a lot of
  prior art and it's not bleeding edge)  is to install the ppa in the
  dockerfile. This is copying the prior art from the Credentials
  dockerfile.

justinhynes commented on Dockerfile:
```diff
@@ -1,4 +1,28 @@
+        language-pack-en \
+        locales \
+        python${PYTHON_VERSION} \
+        python${PYTHON_VERSION}-dev
```
I think this should be good to _install_ Python 3.11, but I'm not sure if it's enough to tell the system to use it.

I was trying to confirm this and realized another issue -- in the Ubuntu base image, `pip` is not installed by default (assuming it must have been in the previous image that came with a preinstalled version of Python). 

When trying to build the Dockerfile, I see:
```
(tubular311) ubuntu@ip-172-31-43-154:~/edx-repos/tubular$ docker build -f Dockerfile -t tubular-upgrade .
[+] Building 86.8s (10/10) FINISHED                                                                                                                                                                                                                                                                            docker:default
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                     0.0s
 => => transferring dockerfile: 1.15kB                                                                                                                                                                                                                                                                                   0.0s
 => [internal] load metadata for docker.io/library/ubuntu:focal                                                                                                                                                                                                                                                          0.1s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                        0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                                                          0.0s
 => [internal] load build context                                                                                                                                                                                                                                                                                        0.2s
 => => transferring context: 4.23MB                                                                                                                                                                                                                                                                                      0.2s
 => CACHED [1/6] FROM docker.io/library/ubuntu:focal@sha256:8feb4d8ca5354def3d8fce243717141ce31e2c428701f6682bd2fafe15388214                                                                                                                                                                                             0.0s
 => [2/6] RUN apt-get update &&     apt-get install -y software-properties-common &&     apt-add-repository -y ppa:deadsnakes/ppa                                                                                                                                                                                       30.9s
 => [3/6] RUN apt-get update && apt-get -qy install --no-install-recommends         build-essential         language-pack-en         locales         python3.11         python3.11-dev                                                                                                                                  55.3s
 => [4/6] WORKDIR /app                                                                                                                                                                                                                                                                                                   0.0s
 => [5/6] ADD . /app                                                                                                                                                                                                                                                                                                     0.1s
 => ERROR [6/6] RUN pip install .                                                                                                                                                                                                                                                                                        0.2s
------
 > [6/6] RUN pip install .:
0.155 /bin/sh: 1: pip: not found
------
Dockerfile:30
--------------------
  28 |     ADD . /app
  29 |
  30 | >>> RUN pip install .
  31 |
--------------------
ERROR: failed to solve: process "/bin/sh -c pip install ." did not complete successfully: exit code: 127
```

I think between lines 25 and 27, we want to add some prior art from our other Dockerfiles:
```
# need to use virtualenv pypi package with Python 3.11
RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python${PYTHON_VERSION}
RUN pip install virtualenv
# create our Python virtual env
ENV VIRTUAL_ENV=/edx/venvs/tubular
RUN virtualenv -p python$PYTHON_VERSION $VIRTUAL_ENV
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

RUN ln -s /usr/bin/python3 /usr/bin/python
```

This should install an appropriate version of pip based on the current Python version specified in the Dockerfile, then will install `virtualenv` and setup a new virtual env, and then creates the link so that we call the expected version of Python.

Edit: You'll also have to use `apt-get` to install curl, so we can pull pip:
```
RUN apt-get update && apt-get -qy install --no-install-recommends \
        build-essential \
        curl \
        language-pack-en \
        locales \
        python${PYTHON_VERSION} \
        python${PYTHON_VERSION}-dev
```

You can then test the Dockerfile build with the command: `docker build -f Dockerfile -t tubular-upgrade .`.

deborahgu committed:
chore: manual pip install

fixing the dockerfile

justinhynes commented on Dockerfile:
```diff
@@ -1,4 +1,36 @@
+# - python${PYTHON_VERSION}-dev; to install header files for python extensions, wheel-building depends on this
+#
+# If you add a package here please include a comment above describing what it is used for
+RUN apt-get update && apt-get -qy install --no-install-recommends \
+        build-essential \
```
We still need to add `curl` here to the list of system requirements.

deborahgu committed:
chore: manual pip install

fixing the dockerfile

deborahgu committed:
chore: manual pip install

fixing the dockerfile

justinhynes commented on Dockerfile:
```diff
@@ -1,6 +1,39 @@
+RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python${PYTHON_VERSION}
+RUN pip install virtualenv
+# create our Python virtual env
+ENV VIRTUAL_ENV=/edx/venvs/credentials
```
One last nit: name of the virtual env is some copy pasta from the credentials dockerfile :-)

justinhynes reviewed (approved):
LGTM, thanks!

deborahgu committed:
chore: manual pip install

fixing the dockerfile
